packages/core/src/scp-cf/destination-accessor.ts

```diff
@@ -482,19 +285,346 @@ const emptyDestinationByType: DestinationsByType = {
+    const providerToken = await DestinationAccessor.getProviderToken(options);
+    const da = new DestinationAccessor(
+      name,
+      options,
+      decodedUserJwt,
+      providerToken
+    );
+
+    return (
+      da.trySubscriberCache() ??
+      (await da.tryProviderCache()) ??
+      (await da.processDestinationResult(await da.tryForOAuth())) ??
+      (await da.processDestinationResult(await da.tryForCert())) ??
+      (await da.processDestinationResult(await da.tryService())) ??
+      undefined
+    );
+  }
+
+  private static async getDecodedUserJwt(
+    options: DestinationOptions & { iss?: string }
+  ): Promise<DecodedJWT | undefined> {
+    return options.userJwt
+      ? verifyJwt(options.userJwt, options)
+      : options.iss
+      ? { iss: options.iss }
+      : undefined;
+  }
+
+  private static async getProviderToken(
+    options: DestinationOptions
+  ): Promise<string> {
+    const { userJwt, ...optionsWithoutJwt } = options;
+    return serviceToken('destination', optionsWithoutJwt);
+  }
+
+  private subscriberDestinationCache: DestinationsByType | undefined;
+  private providerDestinationCache: DestinationsByType | undefined;
+
+  private constructor(
+    private name: string,
+    private options: DestinationOptions,
+    private decodedUserJwt: DecodedJWT | undefined,
+    private providerToken: string
+  ) {
+    this.providerDestinationCache = undefined;
+    this.subscriberDestinationCache = undefined;
+  }
+
+  private get useCache(): boolean {
+    return this.options.useCache || false;
+  }
+
+  private get selectionStrategy(): DestinationSelectionStrategy {
+    return this.options.selectionStrategy
+      ? this.options.selectionStrategy
+      : subscriberFirst;
+  }
+
+  private get isolationStrategy(): IsolationStrategy {
+    return this.options.isolationStrategy
+      ? this.options.isolationStrategy
+      : IsolationStrategy.Tenant;
+  }
+
+  private get decodedProviderToken(): DecodedJWT {
+    return decodeJwt(this.providerToken);
+  }
+
+  private async tryForOAuth(): Promise<Destination | undefined> {
+    const destination = await this.tryService();
+    if (
+      destination &&
+      destination.authentication === 'OAuth2SAMLBearerAssertion'
+    ) {
+      if (destination.systemUser) {
+        const destinationService = getService('destination');
+
+        if (!destinationService) {
+          throw Error(
+            `Failed to fetch destination "${this.name}"! No binding to a destination service found.`
+          );
+        }
+        logger.debug(this.providerToken);
+        return fetchDestination(
+          destinationService.credentials.uri,
+          this.providerToken,
+          this.name,
+          this.options
+        );
+      }
+      if (!this.options.userJwt) {
+        throw Error(
+          `No user token (JWT) has been provided! This is strictly necessary for principal propagation. Value of the JWT: ${this.options.userJwt}.`
```

Comment:
Not sure if adding unsanitized JWTs to error messages is the most secure thing to do. at least the signature could be redacted/falsified.

Reply:
I copy pasted that but I think it does not make sense. If `!this.options.userJwt` there is nothing to print... I just remove it.